### PR TITLE
Enable generated code analysis but analyze razor/cshtml pages only

### PR DESCRIPTION
**Pull request by buyaa-n**
Fixes https://github.com/dotnet/roslyn-analyzers/issues/5971
### CA1416 is disabled for generated code, but we need to enable them for razor/shtml pages
As proposed in https://github.com/dotnet/roslyn/issues/39758: Roslyn already recognizes `<auto-generated />` in a comment on the beginning of source files. We could extend this with an optional type attribute:
```cs
// <auto-generated type="Razor" />
```
We also want to support *.cshtml files, that means the type name might change

- [x] Finalize the type name
- [x] Add CS test
~~- [ ] Add VB test~~ VB not supported
- [x] Apply all code review feedback
- [ ] End to end testing when the source generators start adding the `type="Razor"` attribute to the `<auto-generated/>` comment

~~Creating the PR as a draft until all checks fixed~~ The PR now ready for review

### CODE REVIEW

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs [hunk 1]
```diff
@@ -22,25 +22,6 @@ public partial class PlatformCompatabilityAnalyzerTests
     {
         private const string s_msBuildPlatforms = "build_property._SupportedPlatformList=windows,browser,macOS,maccatalyst, ios, linux;\nbuild_property.TargetFramework=net5.0";
 
-        [Fact(Skip = "TODO need to be fixed: Test for for wrong arguments, not sure how to report the Compiler error diagnostic")]
-        public async Task TestOsPlatformAttributesWithNonStringArgumentAsync()
```
Comment (buyaa-n) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs:26]: This is unrelated, though removing this test as we have an analyzer that explicitly checking the wrong  arguments
https://github.com/dotnet/roslyn-analyzers/blob/main/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/UseValidPlatformString.cs

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs [hunk 2]
```diff
@@ -126,13 +125,15 @@ public sealed partial class PlatformCompatibilityAnalyzer : DiagnosticAnalyzer
             isPortedFxCopRule: false,
             isDataflowRule: false);
 
+        protected abstract bool IsSingleLineComment(SyntaxTrivia trivia);
```
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:128]: We might want to move the language specific code into the `Utilities` instead

File: src/Utilities/Compiler/GeneratedCodeUtilities.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    internal static class GeneratedCodeUtilities
+    {
+        private static readonly string[] s_autoGeneratedStrings = new[] { "<autogenerated", "<auto-generated" };
+
+        internal static bool IsGeneratedCode(SyntaxTree tree, Func<SyntaxTrivia, bool> isComment, CancellationToken cancellationToken)
+        {
+            return BeginsWithAutoGeneratedComment(tree, isComment, cancellationToken);
+        }
+
+        internal static bool IsNotGeneratedCodeOrRazorPage(SyntaxTree tree, Func<SyntaxTrivia, bool> isComment, CancellationToken cancellationToken)
```
Comment (buyaa-n) [inline src/Utilities/Compiler/GeneratedCodeUtilities.cs:16]: Only this method is used for now, could remove unused `IsGeneratedCode(...)` overloads

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs (hunk 1 above)
Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs:26]: Removed all unrelated code

File: src/Utilities/Compiler/GeneratedCodeUtilities.cs (hunk 3 above)
Comment (buyaa-n) [reply src/Utilities/Compiler/GeneratedCodeUtilities.cs]: Removed all unrelated code

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs [hunk 4]
```diff
@@ -126,13 +125,15 @@ public sealed partial class PlatformCompatibilityAnalyzer : DiagnosticAnalyzer
...
+
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(OnlySupportedCsReachable, OnlySupportedCsUnreachable,
             OnlySupportedCsAllPlatforms, SupportedCsAllPlatforms, SupportedCsReachable, UnsupportedCsAllPlatforms, UnsupportedCsReachable);
 
         public override void Initialize(AnalysisContext context)
         {
             context.EnableConcurrentExecution();
-            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:136]: Can you please add a comment here as to why we are analyzing generated code?

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs (hunk 1 above)
Comment (mavasani) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs:26]: Is this deletion related to this PR?

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs [hunk 5]
```diff
@@ -128,6 +109,106 @@ class Target
             await VerifyAnalyzerCSAsync(source, tfmAndOption);
         }
 
+        [Fact]
+        public async Task GeneratedFile_RazorWarns()
```
Comment (mavasani) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs:113]: Just to confirm - did you verify this test fails before your product side changes? The auto-generated comment is not the first line in the file, so it would be good to double check the test failure prior to your changes.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs [hunk 6]
```diff
@@ -128,6 +109,106 @@ class Target
             await VerifyAnalyzerCSAsync(source, tfmAndOption);
         }
 
+        [Fact]
```
Comment (mavasani) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs:112]:
```suggestion
        [Fact, WorkItem(5971, "https://github.com/dotnet/roslyn-analyzers/issues/5971")]
```

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs [hunk 7]
```diff
@@ -128,6 +109,106 @@ class Target
...
+
+#line default
+#line hidden
+    }
+}
+#pragma warning restore 1591";
+            await VerifyAnalyzerCSAsync(csSource, s_msBuildPlatforms);
+        }
+
+        [Fact]
```
Comment (mavasani) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs:162]:
```suggestion
        [Fact, WorkItem(5971, "https://github.com/dotnet/roslyn-analyzers/issues/5971")]
```

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs [hunk 8]
```diff
@@ -128,6 +109,106 @@ class Target
...
+        public async Task GeneratedFile_NotRazorNotWarn()
```
Comment (mavasani) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs:163]: Can we also added VB tests for the same?

File: src/Utilities/Compiler/FileNameUtilities.cs [hunk 9]
```diff
@@ -0,0 +1,144 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System.Diagnostics.CodeAnalysis;
+
+namespace Analyzer.Utilities
+{
+    /// <summary>
+    /// Copied from Roslyn.Utilities.FileNameUtilities.
```
Comment (mavasani) [inline src/Utilities/Compiler/FileNameUtilities.cs:8]: Can we file a tracking issue to remove this code once https://github.com/dotnet/roslyn/issues/39758 is implemented? Also please add a TODO in this comment mentioning the github URI of the new issue filed. Thanks.

Comment (mavasani) [reply src/Utilities/Compiler/FileNameUtilities.cs:8]: Actually, can we just make this a private nested type of GeneratedCodeUtilities, so no one else ends up using these helpers? You can move the above comment and TODO directly to GeneratedCodeUtilities type.

**Review by mavasani — APPROVED** [review]
Please add VB tests and also confirm this does not affect build performance, given this analyzer is on by default on command line.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs (hunk 4 above)
Comment (buyaa-n) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:136]: Sure

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs (hunk 1 above)
Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs:26]: No, it just not needed test

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs (hunk 8 above)
Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs:163]: I have tried to get generated file in VB, but its not supported in blazor or ASP NetCore, though I could just create a fake one and use it

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs (hunk 5 above)
Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs:113]: Do not remember if I did, anyway I can check that

Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs:113]: it fails without changes, and also moving the `auto-generated` comment other rows works

**Comment by buyaa-n** [comment]
Moving to draft until https://github.com/dotnet/roslyn/issues/7578 resolved and the `type="Razor"` attribute added into razor/cshtml pages

**Comment by mavasani** [comment]
@buyaa-n https://github.com/dotnet/roslyn/issues/7578 has been implemented, so we should soon have a public API `IsGeneratedCode` on the various analysis contexts and you wouldn't need to clone the GeneratedCodeUtilities helpers into the repo.

**Comment by jeffhandley** [comment]
Closing out this old draft PR
